### PR TITLE
FeedIteratorInternal

### PR DESCRIPTION
**Pull request by bchong95**
# FeedIteratorInternal

## Description

FeedIterator API did not include any internal methods. This decision was made to allow public users to mock FeedIterator. However the implementation was broken. Internal methods were pushed to all derived classes so there was no inheritance chain. This led to an anti-pattern where you have to try downcasting to all possible types to make it work:

```
        internal bool TryGetContinuationToken(out string state)
        {
            QueryIterator queryIterator = this.feedIterator as QueryIterator;
            if (queryIterator != null)
            {
                return queryIterator.TryGetContinuationToken(out state);
            }

            FeedIteratorCore feedIteratorCore = this.feedIterator as FeedIteratorCore;
            if (feedIteratorCore != null)
            {
                return feedIteratorCore.TryGetContinuationToken(out state);
            }

            throw new ArgumentException($"Unsupported  iterator type of {this.feedIterator.GetType().Name}");
            return this.feedIterator.TryGetContinuationToken(out continuationToken);
        }
```

The solution was to introduce FeedIteratorInternal which inherits from FeedIterator and all the internal types will derive from. That the pattern for TryGetContinuationToken looks like this:

```
        internal bool TryGetContinuationToken(out string state)
        {
            return this.feedIterator.TryGetContinuationToken(out continuationToken);
        }
```

Unfortunately all the resource types (collection, database) need to implement the same pattern to avoid the downcast. Ideally they would all inherit from a resource base class.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs [hunk 1]
```diff
@@ -7,12 +7,9 @@ namespace Microsoft.Azure.Cosmos.Query
     using System;
     using System.Threading;
     using System.Threading.Tasks;
-    using Microsoft.Azure.Cosmos.Handlers;
-    using Microsoft.Azure.Cosmos.Query.Core;
     using Microsoft.Azure.Cosmos.Query.Core.ExecutionContext;
-    using Microsoft.Azure.Cosmos.Query.Core.Monads;
 
-    internal class QueryIterator : FeedIterator
+    internal sealed class QueryIterator : FeedIteratorInternal
```
[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs]: it's internal? who is going to derive from it? A user can inherit from FeedIteratorInternal or Mock<QueryIterator>.

[the remark this answers is not in the document]
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs]: It's internal so no one is going to derive from it. You derive from FeedIteratorInternal or Mock<QueryIterator>.

[the remark this answers is not in the document]
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs]: Friends might need to derive from it or Mock it in a unit test. What benefit do we get from sealing it?

**Comment by sboshra** [comment]
>         internal static string GetContinuationToken(ResponseMessage httpResponseMessage)

public

---
Refers to: Microsoft.Azure.Cosmos/src/FeedIteratorCore.cs:110 in a2d21ba. [](commit_id = a2d21ba2453ef715c354cc1c91f6459e1f2d9d57, deletion_comment = False)

**Pull request merged**